Support types other than `Int` in `randstring`, but convert to `Int` prior to entering the "main" method

Seelengrab opened the pull request:
Fixes #54397

What remains is that the effectively allowed input still needs to be convertible to `Int` and non-negative. So things like `typemax(Int128)` won't work.

Seelengrab committed:
Allow types other than `Int` in `randstring`

Fixes #54397

What remains is that the effectively allowed input still needs
to be convertible to `Int`. So things like `typemax(Int128)` won't
work, and neither will negative values.

KristofferC commented on stdlib/Random/src/misc.jl:
```diff
@@ -71,13 +71,14 @@ let b = UInt8['0':'9';'A':'Z';'a':'z']
     global randstring
 
     function randstring(r::AbstractRNG, chars=b, n::Integer=8)
+        _n = convert(Int, n)
```
I am not sure it is worth doing in this specific instance but presumably structuring it like:

```julia
f(x::Integer) = f(convert(Int, n))
f(x::Int) = ...
```

would reduce compilation time since you would not compile the main function body for multiple integer types.

Seelengrab replied:
For cases where lots of different types are passed to `randstring`, that's certainly the case, yeah. I'm not sure how common that would be though, there's lots of better ways to control the length of a generated string other than through the type.

sjkelly replied:
I think @KristofferC is just referring to the amount of code that is generated by using function barriers or: 
https://docs.julialang.org/en/v1/manual/performance-tips/#kernel-functions

Seelengrab replied:
Yes, I'm referring to the likelihood of hitting that & the additional compilation overhead being a significant slowdown. If this is a bottleneck, I'd first recommend switching to a different scheme for generating the length than optimize the compilation overhead.

sjkelly replied:
The irony appears lost on you, that this PR would allow someone to have a type unstable call site with various `Integer` types for length, rather than error as it currently does.

Seelengrab replied:
Why would that inherently be a problem? We don't prevent type instabilities in user code in other places either.

StefanKarpinski replied:
I think Krostoffer's convert approach is a good idea. Yes, it's ok to just allow a different integer type flow through but in this case Int covers the entire range of reasonable values and converting avoids additional compilation.

Seelengrab replied:
There is no other integer type flowing through though; the code in this PR already calls `_convert(Int, ..)`. This function is a total of 11 lines long, I seriously doubt that this leads to a bottleneck in either compilation time or binary size..

StefanKarpinski replied:
Ok, but every little bit counts. Why not just change it?

Seelengrab replied:
Because I agree with the comment by Kristoffer that it's probably not worth doing in this case. Is it a problem that I agree with the reviewer? As it stands, this just seems like premature optimization of a hypothetical to me.

DilumAluthge replied:
> would reduce compilation time since you would not compile the main function body for multiple integer types.

I have now implemented this.

sjkelly commented:
It is a problem because as this PR is currently written you will generate a
lot of redundant code. You should accept Kristoffer's suggestion for the
signatures as it will improve the quality of this PR.

On Thu, May 9, 2024, 02:07 Sukera ***@***.***> wrote:

> ***@***.**** commented on this pull request.
> ------------------------------
>
> In stdlib/Random/src/misc.jl
> <https://github.com/JuliaLang/julia/pull/54402#discussion_r1595006998>:
>
> > @@ -71,13 +71,14 @@ let b = UInt8['0':'9';'A':'Z';'a':'z']
>      global randstring
>
>      function randstring(r::AbstractRNG, chars=b, n::Integer=8)
> +        _n = convert(Int, n)
>
> Why would that inherently be a problem? We don't prevent type
> instabilities in user code in other places either.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/JuliaLang/julia/pull/54402#discussion_r1595006998>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAP5MDAUQXLRZEXJIYX554LZBMHALAVCNFSM6AAAAABHMSEKMSVHI2DSMVQWIX3LMV43YUDVNRWFEZLROVSXG5CSMV3GSZLXHMZDANBXGIYTEMJQHA>
> .
> You are receiving this because you commented.Message ID:
> ***@***.***>
>

Seelengrab commented:
I'm sure Kristoffer can speak for himself whether this is blocking for the PR. Since he mentioned though that it's probably not worth it in this PR, I've chosen to keep the code simpler to follow without having to disentangle niche dispatch optimizations.

Seelengrab commented:
Does this need anything else (cc @KristofferC)?

nsajko commented on stdlib/Random/src/misc.jl:
```diff
@@ -71,13 +71,14 @@ let b = UInt8['0':'9';'A':'Z';'a':'z']
     global randstring
 
     function randstring(r::AbstractRNG, chars=b, n::Integer=8)
+        _n = convert(Int, n)
```
How about this:

```suggestion
        n = convert(Int, n)::Int
```

Then the rest of the diff can be dropped.

nsajko commented on stdlib/Random/src/misc.jl:
```diff
@@ -71,13 +71,14 @@ let b = UInt8['0':'9';'A':'Z';'a':'z']
             return str
         else
-            v = Vector{T}(undef, n)
+            v = Vector{T}(undef, _n)
```
```suggestion
            v = Vector{T}(undef, n)
```

nsajko commented on stdlib/Random/src/misc.jl:
```diff
@@ -71,13 +71,14 @@ let b = UInt8['0':'9';'A':'Z';'a':'z']
         if T === UInt8
-            str = Base._string_n(n)
-            GC.@preserve str rand!(r, UnsafeView(pointer(str), n), chars)
+            str = Base._string_n(_n)
+            GC.@preserve str rand!(r, UnsafeView(pointer(str), _n), chars)
```
```suggestion
            str = Base._string_n(n)
            GC.@preserve str rand!(r, UnsafeView(pointer(str), n), chars)
```

DilumAluthge committed:
Reduce compilation time (by not compiling the main function body for every Integer type)

As suggested by Kristoffer in https://github.com/JuliaLang/julia/pull/54402#discussion_r1593810608

DilumAluthge committed:
Delineate the different methods more verbosely, by arity

DilumAluthge committed:
Fix a typo

DilumAluthge committed:
Fix another typo

adienes commented:
I do think the original version that converted inside the 3-arg method looked cleaner 😂, but I suppose it's hard to be perfect since we have 3 default positional arguments but want to support basically every possible mix-and-match of calls.

DilumAluthge committed:
Convert one other occurrence of `n::Integer`

Co-authored-by: Andy Dienes <51664769+adienes@users.noreply.github.com>

rfourquet reviewed (approved):
That looks good thank you! Like @adienes , I prefer fewer methods (with less repetition) but it's also fine as is.
